Fgbio/fastqtobam

matthdsm opened the pull request:
<!--
Simplify and generalize module
Added a bunch of extra tests
-->

## PR checklist

Closes #XXX <!-- If this PR fixes an issue, please link it here! -->

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

Midnighter commented on modules/fgbio/fastqtobam/main.nf:
```diff
@@ -9,29 +9,30 @@ process FGBIO_FASTQTOBAM {
         FastqToBam \\
-        -i $reads \\
-        -o "${prefix}_umi_converted.bam" \\
-        --read-structures $read_structure \\
```
Is it fair to remove the `--read-structures` option completely? Are you thinking it should be included via `task.ext.args` when needed?

matthdsm replied:
I'd keep it in `ext.args` (like I did in the tests) since it's a non mandatory option, and the tool also supports straight up conversion without UMI parsing

Midnighter commented on modules/fgbio/fastqtobam/main.nf:
```diff
@@ -9,29 +9,30 @@ process FGBIO_FASTQTOBAM {
-        --sample $meta.id \\
-        --library $meta.id \\
-        $args
+        ${args} \\
```
Would it make sense to include async I/O?

> --async-io[[=true|false]]     Use asynchronous I/O where possible, e.g. for SAM and BAM files. [Default: false]

matthdsm replied:
possibly, but I'd leave that option to `ext.args` since it might not be suitable for all filesystems

Midnighter commented on modules/fgbio/fastqtobam/meta.yml:
```diff
@@ -1,6 +1,6 @@
 name: fgbio_fastqtobam
 description: |
-  Using the FGBIO tools, converts FASTQ files sequenced with UMIs into BAM files, moving the UMI barcode into the RX field of the BAM file
+  Using the FGBIO tools, converts FASTQ files sequenced into BAM files possibly moving the UMI barcode into the RX field of the BAM file
```
```suggestion
  Using the fgbio tools, converts FASTQ files sequenced into BAM files possibly moving the UMI barcode into the RX field of the BAM file
```

Midnighter reviewed (commented):
Thanks for the updates to the module. Looks good to me.

One useful enhancement would be to make the output flexible since fgbio can write cram directly.

matthdsm commented:
> One useful enhancement would be to make the output flexible since fgbio can write cram directly.

It should support cram, but I didn't find any options in the [manual](http://fulcrumgenomics.github.io/fgbio/tools/latest/FastqToBam.html) how to enable this feature.

matthdsm commented:
How would you suggest we add the option? We could add a flag, but that just seems a bit ugly to me

matthdsm commented:
Personally, I like we've got the option to go straight to cram, but I'd still prefer using samtools for compression since we've got more control over the settings

Midnighter commented:
> How would you suggest we add the option? We could add a flag, but that just seems a bit ugly to me

Yeah, not entirely sure about that either. Maybe we could expect users to change the `task.ext.prefix` to include `.cram`? And otherwise default to creating `"${prefix}.bam"`?

> Personally, I like we've got the option to go straight to cram, but I'd still prefer using samtools for compression since we've got more control over the settings

What kind of settings are you thinking of? More control is usually nice but submitting less jobs and moving less files around is also nice.

matthdsm commented:
> And otherwise default to creating "${prefix}.bam"?
That's a possiblity, though it seems a but hacky

> What kind of settings are you thinking of?
Compression factors, block sizes etc. You can control all of those through samtools, but I don't see any options available here. There might be the possibility to tweak some HTSjdk settings, but those aren't documented

FriederikeHanssen commented on modules/fgbio/fastqtobam/main.nf:
```diff
@@ -9,29 +9,30 @@ process FGBIO_FASTQTOBAM {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def sample_name = args.contains("--sample") ? "" : "--sample ${prefix}"
+    def library_name = args.contains("--library") ? "" : "--library ${prefix}"
```
hm intuitively I would say they should go to args, but I  have been seeing some modules that have been more lenient on this :D

matthdsm replied:
Both those options are mandatory, so that's why they're defined in the module, with a fallback to `ext.args`

Midnighter commented on modules/fgbio/fastqtobam/main.nf:
```diff
@@ -22,15 +22,17 @@ process FGBIO_FASTQTOBAM {
     def sample_name = args.contains("--sample") ? "" : "--sample ${prefix}"
     def library_name = args.contains("--library") ? "" : "--library ${prefix}"
-
+    def output = prefix.endsWith(".bam") ? ${prefix} :
+                 prefix.endsWith(".cram")? ${prefix} :
+                 "${prefix}.bam"
```
Maybe a bit slimmer with a regex (untested):

```suggestion
    def output = prefix =~ /\.(bam|cram)$/ ? prefix : "${prefix}.bam"
```

matthdsm commented:
I added some updates, but it seems like the config options aren't getting picked up. Probably something wrong with the selectors.
Suggestions are welcome!

Midnighter commented:
> I added some updates, but it seems like the config options aren't getting picked up. Probably something wrong with the selectors. Suggestions are welcome!

Can you say more about what you mean exactly? Selectors are usually a workflow configuration thing and not for modules or am I wrong?

matthdsm commented:
I'm talking about `nextflow.config` in the module tests. I'm trying to add different config options to test for multiple usecases.

matthdsm commented:
I dropped the checksum for the cram output because the tests don't seem able to get a match.
However:
- I converted 2 separate runs to sam and ran a `diff` without anything coming up
- I manually checksummed 2 separate runs and came up with the samen checksum
So no idea what's going wrong here...

Midnighter commented on modules/fgbio/fastqtobam/main.nf:
```diff
@@ -9,29 +9,31 @@ process FGBIO_FASTQTOBAM {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    def sample_name = args.contains("--sample") ? "" : "--sample ${prefix}"
+    def library_name = args.contains("--library") ? "" : "--library ${prefix}"
```
I see an issue with these, when the prefix is already named `.bam` or `.cram` then this will also be the sample and library name. I think that should either be stripped or use an optional input to toggle CRAM. What do you think?

matthdsm replied:
I noticed that too, but when you're already adding option to modify the `$prefix` you might as well set `--sample` and `--library` in `ext.args` IMO.

matthdsm replied:
I added a test where both are set explicitly to test for this case

Midnighter replied:
Perfect, I only commented on this because you were _not_ configuring it that way in the tests 😉

matthdsm replied:
yeah, I did use separate tests. I probably should've done it all in one.

Midnighter reviewed (approved):
Very good improvements. It's slightly uncomfortable that the cram md5sums don't match or can otherwise be compared but overall the module looks much better.

matthdsm commented:
Thanks! 
It's kind of weird the checksums don't match, but I did as much as possible to rule out any weird stuff 🤷🏻‍♂️
